File: sdk/cs/src/OpenAI/EmbeddingClient.cs
// --------------------------------------------------------------------------------------------------------------------
// <copyright company="Microsoft">
//   Copyright (c) Microsoft. All rights reserved.
// </copyright>
// --------------------------------------------------------------------------------------------------------------------

namespace Microsoft.AI.Foundry.Local;

using Betalgo.Ranul.OpenAI.ObjectModels.ResponseModels;

using Microsoft.AI.Foundry.Local.Detail;
using Microsoft.AI.Foundry.Local.OpenAI;
using Microsoft.Extensions.Logging;

/// <summary>
/// Embedding Client that uses the OpenAI API.
/// Implemented using Betalgo.Ranul.OpenAI SDK types.
/// </summary>
public class OpenAIEmbeddingClient
{
    private readonly string _modelId;

    private readonly ICoreInterop _coreInterop = FoundryLocalManager.Instance.CoreInterop;
    private readonly ILogger _logger = FoundryLocalManager.Instance.Logger;

    internal OpenAIEmbeddingClient(string modelId)
    {
        _modelId = modelId;
    }

    /// <summary>
    /// Settings that are supported by Foundry Local for embeddings.
    /// </summary>
    public record EmbeddingSettings
    {
        /// <summary>
        /// The number of dimensions the resulting output embeddings should have.
        /// </summary>
        public int? Dimensions { get; set; }

        /// <summary>
        /// The format to return the embeddings in. Can be either "float" or "base64".
        /// </summary>
        public string? EncodingFormat { get; set; }
    }

    /// <summary>
    /// Settings to use for embedding requests using this client.
    /// </summary>
    public EmbeddingSettings Settings { get; } = new();

    /// <summary>
    /// Generate embeddings for the given input text.
    /// </summary>
    /// <param name="input">The text to generate embeddings for.</param>
    /// <param name="ct">Optional cancellation token.</param>
    /// <returns>Embedding response containing the embedding vector.</returns>
    public async Task<EmbeddingCreateResponse> GenerateEmbeddingAsync(string input,
                                                                      CancellationToken? ct = null)
    {
        return await Utils.CallWithExceptionHandling(
            () => GenerateEmbeddingImplAsync(input, ct),
            "Error during embedding generation.", _logger).ConfigureAwait(false);
    }

    /// <summary>
    /// Generate embeddings for multiple input texts in a single request.
    /// </summary>
    /// <param name="inputs">The texts to generate embeddings for.</param>
    /// <param name="ct">Optional cancellation token.</param>
    /// <returns>Embedding response containing one embedding vector per input.</returns>
    public async Task<EmbeddingCreateResponse> GenerateEmbeddingsAsync(IEnumerable<string> inputs,
                                                                       CancellationToken? ct = null)
    {
        return await Utils.CallWithExceptionHandling(
            () => GenerateEmbeddingsImplAsync(inputs, ct),
            "Error during batch embedding generation.", _logger).ConfigureAwait(false);
    }

    private async Task<EmbeddingCreateResponse> GenerateEmbeddingImplAsync(string input,
                                                                            CancellationToken? ct)
    {
        var embeddingRequest = EmbeddingCreateRequestExtended.FromUserInput(_modelId, input, Settings);
        var embeddingRequestJson = embeddingRequest.ToJson();

        var request = new CoreInteropRequest { Params = new() { { "OpenAICreateRequest", embeddingRequestJson } } };
        var response = await _coreInterop.ExecuteCommandAsync("embeddings", request,
                                                                ct ?? CancellationToken.None).ConfigureAwait(false);

        return response.ToEmbeddingResponse(_logger);
    }

    private async Task<EmbeddingCreateResponse> GenerateEmbeddingsImplAsync(IEnumerable<string> inputs,
                                                                             CancellationToken? ct)
    {
        var embeddingRequest = EmbeddingCreateRequestExtended.FromUserInput(_modelId, inputs, Settings);
        var embeddingRequestJson = embeddingRequest.ToJson();

        var request = new CoreInteropRequest { Params = new() { { "OpenAICreateRequest", embeddingRequestJson } } };
        var response = await _coreInterop.ExecuteCommandAsync("embeddings", request,
                                                                ct ?? CancellationToken.None).ConfigureAwait(false);

        return response.ToEmbeddingResponse(_logger);
    }
}

File: sdk/rust/src/openai/embedding_client.rs
//! OpenAI-compatible embedding client.

use std::sync::Arc;

use serde_json::{json, Value};

use crate::detail::core_interop::CoreInterop;
use crate::error::Result;

/// OpenAI-compatible embedding response.
#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
pub struct EmbeddingData {
    /// The index of the embedding in the list.
    pub index: i32,
    /// The embedding vector.
    pub embedding: Vec<f64>,
}

/// OpenAI-compatible embedding response.
#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
pub struct EmbeddingResponse {
    /// The model used for generation.
    pub model: String,
    /// The object type (always "list").
    pub object: Option<String>,
    /// List of embedding results.
    pub data: Vec<EmbeddingData>,
}

/// Tuning knobs for embedding requests.
///
/// Use the chainable setter methods to configure, e.g.:
///
/// ```ignore
/// let client = model.create_embedding_client()
///     .dimensions(512);
/// ```
#[derive(Debug, Clone, Default)]
pub struct EmbeddingClientSettings {
    dimensions: Option<u32>,
    encoding_format: Option<String>,
}

impl EmbeddingClientSettings {
    fn serialize(&self, model_id: &str, input: &str) -> Value {
        let mut map = serde_json::Map::new();

        map.insert("model".into(), json!(model_id));
        map.insert("input".into(), json!(input));

        if let Some(dims) = self.dimensions {
            map.insert("dimensions".into(), json!(dims));
        }
        if let Some(ref fmt) = self.encoding_format {
            map.insert("encoding_format".into(), json!(fmt));
        }

        Value::Object(map)
    }
}

/// Client for OpenAI-compatible embedding generation backed by a local model.
pub struct EmbeddingClient {
    model_id: String,
    core: Arc<CoreInterop>,
    settings: EmbeddingClientSettings,
}

impl EmbeddingClient {
    pub(crate) fn new(model_id: &str, core: Arc<CoreInterop>) -> Self {
        Self {
            model_id: model_id.to_owned(),
            core,
            settings: EmbeddingClientSettings::default(),
        }
    }

    /// Set the number of dimensions for the output embeddings.
    pub fn dimensions(mut self, v: u32) -> Self {
        self.settings.dimensions = Some(v);
        self
    }

    /// Set the encoding format ("float" or "base64").
    pub fn encoding_format(mut self, v: impl Into<String>) -> Self {
        self.settings.encoding_format = Some(v.into());
        self
    }

    /// Generate embeddings for the given input text.
    pub async fn generate_embedding(&self, input: &str) -> Result<EmbeddingResponse> {
        Self::validate_input(input)?;

        let request = self.settings.serialize(&self.model_id, input);
        let params = json!({
            "Params": {
                "OpenAICreateRequest": serde_json::to_string(&request)?
            }
        });

        let raw = self
            .core
            .execute_command_async("embeddings".into(), Some(params))
            .await?;
        let parsed: EmbeddingResponse = serde_json::from_str(&raw)?;
        Ok(parsed)
    }

    fn validate_input(input: &str) -> Result<()> {
        if input.trim().is_empty() {
            return Err(crate::error::FoundryLocalError::Validation {
                reason: "input must be a non-empty string".into(),
            });
        }
        Ok(())
    }
}
